src/Compilers/CSharp/Test/Symbol/Symbols/FunctionPointerTypeSymbolTests.cs

```diff
@@ -692,5 +698,92 @@ .maxstack  8
         }
+
+        [Fact]
+        public void RequiresUnsafeInSignature()
+        {
+            var comp = CreateFunctionPointerCompilation(@"
+class C
+{
+    delegate*<void> _field;
+    delegate*<void> Property { get; set; }
+    delegate*<void> M(delegate*<void> param)
+    {
+        delegate*<void> local1;
+        /**/delegate/**/*/**/<void> local2;
+        throw null;
+    }
+}");
+
+            comp.VerifyDiagnostics(
+                    // (4,5): error CS0214: Pointers and fixed size buffers may only be used in an unsafe context
+                    //     delegate*<void> _field;
+                    Diagnostic(ErrorCode.ERR_UnsafeNeeded, "delegate*").WithLocation(4, 5),
+                    // (4,21): warning CS0169: The field 'C._field' is never used
+                    //     delegate*<void> _field;
+                    Diagnostic(ErrorCode.WRN_UnreferencedField, "_field").WithArguments("C._field").WithLocation(4, 21),
+                    // (5,5): error CS0214: Pointers and fixed size buffers may only be used in an unsafe context
+                    //     delegate*<void> Property { get; set; }
+                    Diagnostic(ErrorCode.ERR_UnsafeNeeded, "delegate*").WithLocation(5, 5),
+                    // (6,5): error CS0214: Pointers and fixed size buffers may only be used in an unsafe context
+                    //     delegate*<void> M(delegate*<void> param)
+                    Diagnostic(ErrorCode.ERR_UnsafeNeeded, "delegate*").WithLocation(6, 5),
+                    // (6,23): error CS0214: Pointers and fixed size buffers may only be used in an unsafe context
+                    //     delegate*<void> M(delegate*<void> param)
+                    Diagnostic(ErrorCode.ERR_UnsafeNeeded, "delegate*").WithLocation(6, 23),
+                    // (8,9): error CS0214: Pointers and fixed size buffers may only be used in an unsafe context
+                    //         delegate*<void> local1;
+                    Diagnostic(ErrorCode.ERR_UnsafeNeeded, "delegate*").WithLocation(8, 9),
+                    // (8,25): warning CS0168: The variable 'local1' is declared but never used
+                    //         delegate*<void> local1;
+                    Diagnostic(ErrorCode.WRN_UnreferencedVar, "local1").WithArguments("local1").WithLocation(8, 25),
+                    // (9,13): error CS0214: Pointers and fixed size buffers may only be used in an unsafe context
+                    //         /**/delegate/**/*/**/<void> local2;
+                    Diagnostic(ErrorCode.ERR_UnsafeNeeded, "delegate/**/*").WithLocation(9, 13),
+                    // (9,37): warning CS0168: The variable 'local1' is declared but never used
+                    //         /**/delegate/**/*/**/<void> local2;
+                    Diagnostic(ErrorCode.WRN_UnreferencedVar, "local2").WithArguments("local2").WithLocation(9, 37));
+        }
+
+        [Fact]
+        public void MisdeclaredArraysWithLocalDeclarationsAreHandled()
```

Comment:
>MisdeclaredArraysWithLocalDeclarationsAreHandled [](start = 20, length = 48)

It would be good to add tests covering other code paths you adjusted.